
### Lab 02

23 . febrero . 2026

En este laboratorio vamos a implementar varios métodos de aprendizaje supervisado y aplicarlos a conjuntos de datos.



In [7]:
import numpy as np
import pandas as pd
from sklearn.naive_bayes import CategoricalNB
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score

---
### 1. 
Para los siguientes datos: Construir el clasificador bayesiano óptimo. Indicar la regla de clasificación. Hallar el error de este clasificador.

|       | X = 1 | X = 2 | X = 3 | X = 4 |
|-------|-------|-------|-------|-------|
| Y = 0 | 0.09  | 0.16  | 0.17  | 0.05  |
| Y = 1 | 0.15  | 0.07  | 0.10  | 0.21  |


In [ ]:
# Definir los datos de la tabla de probabilidad conjunta P(X, Y)
# Las filas son Y=0, Y=1 y las columnas son X=1, X=2, X=3, X=4
datos = np.array([
    [0.09, 0.16, 0.17, 0.05], # Y = 0
    [0.15, 0.07, 0.10, 0.21]  # Y = 1
])

columnas = ['X=1', 'X=2', 'X=3', 'X=4']
indices = ['Y=0', 'Y=1']
df = pd.DataFrame(datos, columns=columnas, index=indices)

print("Tabla de Probabilidades Conjuntas P(X, Y):")
display(df) 


regla_clasificacion = df.idxmax(axis=0)

print("\n--- Regla de Clasificación ---")
for x in columnas:
    clase_predicha = regla_clasificacion[x]
    print(f"Para {x} -> Predicción: {clase_predicha} (Prob: {df.loc[clase_predicha, x]:.2f})")

# Hallar el error del clasificador
# El error es la suma de las probabilidades de las clases NO elegidas 
probabilidades_error = df.min(axis=0)
error_total = probabilidades_error.sum()

print("\n--- Error del Clasificador ---")
print(f"Probabilidades de error por cada X:\n{probabilidades_error.to_string()}")
print(f"\nError total del clasificador: {error_total:.2f}")

Tabla de Probabilidades Conjuntas P(X, Y):


,X=1,X=2,X=3,X=4
Y=0,0.09,0.16,0.17,0.05
Y=1,0.15,0.07,0.10,0.21



--- Regla de Clasificación ---
Para X=1 -> Predicción: Y=1 (Prob: 0.15)
Para X=2 -> Predicción: Y=0 (Prob: 0.16)
Para X=3 -> Predicción: Y=0 (Prob: 0.17)
Para X=4 -> Predicción: Y=1 (Prob: 0.21)

--- Error del Clasificador ---
Probabilidades de error por cada X:
X=1    0.09
X=2    0.07
X=3    0.10
X=4    0.05

Error total del clasificador: 0.31


El error del clasificador bayesiano óptimo es:

$$
\boxed{P(\text{error}) = 0.31}
$$

Por lo tanto, la **accuracy máxima posible** para este problema es:

$$
1 - 0.31 = 0.69
$$


---
### 2.

Construir un clasificador bayesiano óptimo para dos distribuciones continuas con densidad exponencial  
$f_0 \sim \mathrm{Exp}(1)$ y $f_1 \sim \mathrm{Exp}(3)$.

Esto es, distribuciones con densidades condicionales

$$
f_0(x) = f(x \mid Y = 0) = e^{-x}
$$

$$
f_1(x) = f(x \mid Y = 1) = 3e^{-3x}
$$

En cada caso, construir la regla de clasificación, las regiones de clasificación y calcular el error del clasificador


---
### 3.

La siguiente tabla de datos muestra información de la decisión de jugar o no jugar golf, en función de varios factores como el tipo de clima, temperatura, humedad, o si hay viento, para 14 observaciones diferentes.

| Obs | Outlook   | Temperature | Humidity | Windy | Play Golf? |
|-----|-----------|------------|----------|-------|------------|
| 0   | Rainy     | Hot        | High     | False | No         |
| 1   | Rainy     | Hot        | High     | True  | No         |
| 2   | Overcast  | Hot        | High     | False | Yes        |
| 3   | Sunny     | Mild       | High     | False | Yes        |
| 4   | Sunny     | Cool       | Normal   | False | Yes        |
| 5   | Sunny     | Cool       | Normal   | True  | No         |
| 6   | Overcast  | Cool       | Normal   | True  | Yes        |
| 7   | Rainy     | Mild       | High     | False | No         |
| 8   | Rainy     | Cool       | Normal   | False | Yes        |
| 9   | Sunny     | Mild       | Normal   | False | Yes        |
| 10  | Rainy     | Mild       | Normal   | True  | Yes        |
| 11  | Overcast  | Mild       | High     | True  | Yes        |
| 12  | Overcast  | Hot        | Normal   | False | Yes        |
| 13  | Sunny     | Mild       | High     | True  | No         |

Las variables son:

- $X_1$ = Outlook  
- $X_2$ = Temperature  
- $X_3$ = Humidity  
- $X_4$ = Windy  

Con base en la evidencia de la tabla, construya **dos** de los siguientes clasificadores:

- (a) KNN  
- (b) Naive Bayes  
- (c) Regresión logística  

para determinar si se debe o no jugar golf en los siguientes casos:

- $x = (\text{Rainy}, \text{Hot}, \text{High}, \text{False})$
- $x = (\text{Sunny}, \text{Hot}, \text{Normal}, \text{False})$



Compare el desempeño de sus dos clasificadores e indique con cuál se obtienen mejores resultados.  
Utilice métricas como:

- Accuracy  
- Precision  
- Recall  
- $F_1$ Score  
- ROC AUC  

para validar su argumento.




In [4]:
# 1. Recreando la tabla de datos de 14 observaciones
data = {
    'Outlook': ['Rainy', 'Rainy', 'Overcast', 'Sunny', 'Sunny', 'Sunny', 'Overcast', 'Rainy', 'Rainy', 'Sunny', 'Rainy', 'Overcast', 'Overcast', 'Sunny'],
    'Temperature': ['Hot', 'Hot', 'Hot', 'Mild', 'Cool', 'Cool', 'Cool', 'Mild', 'Cool', 'Mild', 'Mild', 'Mild', 'Hot', 'Mild'],
    'Humidity': ['High', 'High', 'High', 'High', 'Normal', 'Normal', 'Normal', 'High', 'Normal', 'Normal', 'Normal', 'High', 'Normal', 'High'],
    'Windy': [False, True, False, False, False, True, True, False, False, False, True, True, False, True],
    'Play_Golf': ['No', 'No', 'Yes', 'Yes', 'Yes', 'No', 'Yes', 'No', 'Yes', 'Yes', 'Yes', 'Yes', 'Yes', 'No']
}
df = pd.DataFrame(data)

# 2. Transformar variables usando one-hot encoding
X = pd.get_dummies(df[['Outlook', 'Temperature', 'Humidity', 'Windy']], dtype=int)
y = df['Play_Golf'].map({'Yes': 1, 'No': 0})

# 3. Preparar los dos casos específicos a predecir
casos_prueba = pd.DataFrame({
    'Outlook': ['Rainy', 'Sunny'],
    'Temperature': ['Hot', 'Hot'],
    'Humidity': ['High', 'Normal'],
    'Windy': [False, False]
})

# Aplicar get_dummies y alinear las columnas con X para evitar errores de dimensión
X_prueba = pd.get_dummies(casos_prueba, dtype=int)
X_prueba = X_prueba.reindex(columns=X.columns, fill_value=0)

Clasificador 1: Naïve Bayes

¿Por qué lo escogimos?
1. Manejo de datos categóricos: Todas las variables independientes (Outlook, Temperature, Humidity, Windy) son puramente categóricas. Naïve Bayes es excepcionalmente bueno trabajando con este tipo de atributos calculando probabilidades condicionales directas.
2. Escasez de datos: La tabla solo cuenta con 14 observaciones. Naïve Bayes no requiere de grandes volúmenes de datos para estimar los parámetros necesarios para la clasificación (las probabilidades a priori y condicionales), lo que ayuda a responder la pregunta final del laboratorio sobre cómo construir modelos con pocos datos.

In [5]:
# Entrenar el modelo
nb_model = CategoricalNB()
nb_model.fit(X, y)

# Calcular predicciones para el set de datos
y_pred_nb = nb_model.predict(X)
y_prob_nb = nb_model.predict_proba(X)[:, 1]
pred_casos_nb = nb_model.predict(X_prueba)

# Mostrar Métricas
print("--- MÉTRICAS NAÏVE BAYES ---")
print(f"Accuracy:  {accuracy_score(y, y_pred_nb):.4f}")
print(f"Precision: {precision_score(y, y_pred_nb):.4f}")
print(f"Recall:    {recall_score(y, y_pred_nb):.4f}")
print(f"F1 Score:  {f1_score(y, y_pred_nb):.4f}")
print(f"ROC AUC:   {roc_auc_score(y, y_prob_nb):.4f}")

# Mostrar Predicciones solicitadas
print("\n--- PREDICCIONES DE LOS CASOS ---")
print(f"x = (Rainy, Hot, High, False) -> ¿Play Golf?: {'Yes' if pred_casos_nb[0]==1 else 'No'}")
print(f"x = (Sunny, Hot, Normal, False) -> ¿Play Golf?: {'Yes' if pred_casos_nb[1]==1 else 'No'}")

--- MÉTRICAS NAÏVE BAYES ---
Accuracy:  0.8571
Precision: 0.8889
Recall:    0.8889
F1 Score:  0.8889
ROC AUC:   0.8889

--- PREDICCIONES DE LOS CASOS ---
x = (Rainy, Hot, High, False) -> ¿Play Golf?: No
x = (Sunny, Hot, Normal, False) -> ¿Play Golf?: Yes


Clasificador 2: Regresión Logística

¿Por qué lo escogimos?
1. Naturaleza binaria del problema: La variable objetivo "Play Golf?" es estrictamente binaria (Yes/No), y la regresión logística está diseñada matemáticamente y optimizada exactamente para este tipo de salidas.
2. Evaluación con ROC AUC: El laboratorio pide comparar los modelos utilizando métricas como ROC AUC. La Regresión Logística es un modelo probabilístico que por naturaleza devuelve una probabilidad continua entre 0 y 1 antes de aplicar el umbral de decisión, lo que hace que trazar y evaluar la curva ROC sea muy natural y preciso en comparación con KNN, el cual puede ser más rígido con las probabilidades.

In [6]:
# Entrenar el modelo
lr_model = LogisticRegression(random_state=42)
lr_model.fit(X, y)

# Calcular predicciones para el set de datos
y_pred_lr = lr_model.predict(X)
y_prob_lr = lr_model.predict_proba(X)[:, 1]
pred_casos_lr = lr_model.predict(X_prueba)

# Mostrar Métricas
print("--- MÉTRICAS REGRESIÓN LOGÍSTICA ---")
print(f"Accuracy:  {accuracy_score(y, y_pred_lr):.4f}")
print(f"Precision: {precision_score(y, y_pred_lr):.4f}")
print(f"Recall:    {recall_score(y, y_pred_lr):.4f}")
print(f"F1 Score:  {f1_score(y, y_pred_lr):.4f}")
print(f"ROC AUC:   {roc_auc_score(y, y_prob_lr):.4f}")

# Mostrar Predicciones solicitadas
print("\n--- PREDICCIONES DE LOS CASOS ---")
print(f"x = (Rainy, Hot, High, False) -> ¿Play Golf?: {'Yes' if pred_casos_lr[0]==1 else 'No'}")
print(f"x = (Sunny, Hot, Normal, False) -> ¿Play Golf?: {'Yes' if pred_casos_lr[1]==1 else 'No'}")

--- MÉTRICAS REGRESIÓN LOGÍSTICA ---
Accuracy:  0.8571
Precision: 0.8182
Recall:    1.0000
F1 Score:  0.9000
ROC AUC:   0.9778

--- PREDICCIONES DE LOS CASOS ---
x = (Rainy, Hot, High, False) -> ¿Play Golf?: No
x = (Sunny, Hot, Normal, False) -> ¿Play Golf?: Yes


### Resultados de Predicción
Ambos modelos coinciden en los casos de prueba:
1.  `(Rainy, Hot, High, False)` → **No Jugar**
2.  `(Sunny, Hot, Normal, False)` → **Sí Jugar**

### Conclusiones
La **Regresión Logística** resulta ser el modelo superior para este caso, ya que alcanza un **ROC AUC de 0.9778** y un **Recall de 1.0**, demostrando una capacidad casi perfecta para separar las clases y no omitir ningún caso positivo ("Yes").



Responda la siguiente pregunta:

**¿Cómo construir un clasificador bueno, con tan pocos datos?**


Para construir modelos confiables con muestras mínimas (N=14):
1.  **Simplicidad:** Usar modelos con "fuerte sesgo" (como los probados) que evitan el sobreajuste (overfitting).
2.  **Codificación:** Transformar variables categóricas correctamente (One-Hot Encoding).
3.  **Métricas Robustas:** No fiarse solo del Accuracy; priorizar F1-Score y ROC AUC para entender la capacidad real de distinción del modelo.


---
### 4.

Construir un clasificador binario para el siguiente conjunto de datos de **Spam / No Spam**:

https://www.kaggle.com/datasets/alexandercarreo/es-spamham-dataset-para-deteccion-de-spam?select=spam_ham_es_dataset_v1.csv

Usted puede elegir cualquiera de los algoritmos vistos en clase para construir su clasificador, pero el objetivo es utilizar aquel con mejor desempeño (menor error) posible.

Utilice métricas como:

- Accuracy  
- Precision  
- Recall  
- $F_1$ Score  
- ROC AUC  
- Binary Crossentropy  

para validar el desempeño de su clasificador.


Para ello, cuando su algoritmo así lo requiera, deberá construir un *embedding* adecuado de sus textos a un vector numérico.

### Sugerencias:

- No olvide hacer un uso adecuado del conjunto de entrenamiento, el conjunto de validación y el conjunto de prueba.
- Uso de técnicas de validación cruzada para una buena medición del desempeño.
- Se sugiere el uso de técnicas de procesamiento de lenguaje natural (NLP). Investigue en qué consiste y para qué sirven los siguientes conceptos, y aplíquelos en su clasificador si lo considera conveniente:

  - Remoción de *stopwords*  
  - Uso de *lemmatization*  
  - Uso de vectorizaciones:
    - Bag of Words  
    - TF-IDF  
    - Doc2Vec
